### PR TITLE
RavenDB-16727 Curl Command Dropdown (Import & Export)

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

File: src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts [hunk 1]
```diff
@@ -124,8 +130,28 @@ class exportDatabase extends viewModelBase {
 
             return collections.filter(x => x.toLowerCase().includes(filterLowerCase));
         });
+        
+        const commandEndpointUrl = (db: database) => appUrl.forServer() + appUrl.forDatabaseQuery(db) + endpoints.databases.smuggler.smugglerExport;
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts]: can we extract shared logic into new method and only provide some common escape function ?

File: src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts [hunk 2]
```diff
@@ -135,17 +161,48 @@ class exportDatabase extends viewModelBase {
...
+            }
+            
+            const fileName = args.FileName;
+            delete args.FileName;
+            const json = JSON.stringify(args);
+
+            return `curl -o '${fileName}.ravendbdump' --data DownloadOptions='${json}' ${commandEndpointUrl(db)}`;
+        });
+
+        this.effectiveCommandLabel = ko.pureComputed(() => {
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts]: remove this and simply use: getCommandTypeLabel(this.effectiveCommand) in view.

File: src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts [hunk 3]
```diff
@@ -239,6 +296,15 @@ class exportDatabase extends viewModelBase {
                     }).always(() => exportDatabase.isExporting(false));
             });
     }
+
+    getCommandTypeLabel(cmdType: commandLineType) {
+        return `Export Command - ${cmdType}`;
+    }
+
+    copyCommandToClipboard() {
+        let command = this.effectiveCommand();
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts]: const

File: src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts [hunk 4]
```diff
@@ -116,7 +120,24 @@ class importDatabaseFromFile extends viewModelBase {
...
+            return `curl -F 'importOptions=${json}' -F 'file=@${fileName}' ${commandEndpointUrl(db)}`;
+        });
+
+        this.effectiveCommandLabel = ko.pureComputed(() => {
+            const cmdType = this.effectiveCommandType();
+            return this.getCommandTypeLabel(cmdType);
+        });
+        
+        this.effectiveCommand = ko.pureComputed(() => {
+            const cmdType = this.effectiveCommandType();
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts]: apply same comments to import like I wrote in export.

File: src/Raven.Studio/typings/_studio/dto.d.ts [hunk 5]
```diff
@@ -745,3 +745,38 @@ interface sampleCode {
...
+    nodeTag: string;
+    hideDatabaseName: boolean;
+    even: boolean;
+    noData: boolean;
+}
+
+interface cachedDateValue<T> {
+    date: Date;
+    value: T;
+}
```
Comment (ml054) [inline src/Raven.Studio/typings/_studio/dto.d.ts]: what's that?

File: src/Raven.Studio/typings/_studio/dto.d.ts [hunk 6]
```diff
@@ -745,3 +745,38 @@ interface sampleCode {
...
+
+type widgetType = Raven.Server.Dashboard.Cluster.ClusterDashboardNotificationType | "Welcome" | "License";
+
+type databaseAccessLevel = `Database${Raven.Client.ServerWide.Operations.Certificates.DatabaseAccess}`;
+type securityClearance = Raven.Client.ServerWide.Operations.Certificates.SecurityClearance;
+type accessLevel = databaseAccessLevel | securityClearance;
```
Comment (ml054) [inline src/Raven.Studio/typings/_studio/dto.d.ts]: what's that?

File: src/Raven.Studio/wwwroot/App/views/database/tasks/exportDatabase.html [hunk 7]
```diff
@@ -172,14 +172,24 @@ <h4>Select collections:</h4>
...
-                        </span>
-                            <input onClick="this.select();" class="form-control" readonly data-bind="value: exportCommand" />
+                    <div class="flex-horizontal" data-bind="with: $root">
+                        <div class="btn-group">
+                            <button type="button" class="btn btn-default dropdown-toggle" data-toggle="dropdown" aria-expanded="false">
+                                <span data-bind="text: effectiveCommandLabel"></span>
+                                <span class="caret margin-left margin-left-xs"></span>
+                            </button>
+                            <ul class="dropdown-menu">
+                                <li><a href="#" data-bind="click: effectiveCommandType.bind(effectiveCommandType, 'PowerShell')"><span data-bind="text: getCommandTypeLabel('PowerShell')"></span></a></li>
```
Comment (ml054) [inline src/Raven.Studio/wwwroot/App/views/database/tasks/exportDatabase.html]: let's expose list of command types in model and iterate though them - code will be more bullet proof.

File: src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html [hunk 8]
```diff
@@ -180,33 +180,25 @@ <h3>Advanced</h3>
...
-                                    <a href="#" data-bind="click: _.partial($root.copyCommandToClipboard, $root.importCommandCmd())">
-                                        <i class="icon-copy"></i> Copy as cURL (cmd)
-                                    </a>
-                                </li>
-                                <li>
-                                    <a href="#" data-bind="click: _.partial($root.copyCommandToClipboard, $root.importCommandBash())">
-                                        <i class="icon-copy"></i> Copy as cURL (bash)
-                                    </a>
-                                </li>
+                                <li><a href="#" data-bind="click: effectiveCommandType.bind(effectiveCommandType, 'PowerShell')"><span data-bind="text: getCommandTypeLabel('PowerShell')"></span></a></li>
```
Comment (ml054) [inline src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html]: same here

File: src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html [hunk 9]
```diff
@@ -180,33 +180,25 @@ <h3>Advanced</h3>
...
+                                <li><a href="#" data-bind="click: effectiveCommandType.bind(effectiveCommandType, 'Bash')"><span data-bind="text: getCommandTypeLabel('Bash')"></span></a></li>
                             </ul>
                         </div>
                         <div class="flex-grow">
-                            <input onClick="this.select();" class="form-control" readonly data-bind="value: $root.importCommandPowerShell" />
+                            <input onClick="this.select();" class="form-control" readonly data-bind="value: effectiveCommand" />
                         </div>
-                    </div>    
+                        <button class="btn btn-default" data-bind="click: copyCommandToClipboard" title="Copy import command to clipboard">
+                            <i class="icon-copy-to-clipboard"></i><span>Copy</span>
```
Comment (ml054) [inline src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html]: I wonder if we need icon + text, or only icon?

File: src/Raven.Studio/typings/_studio/dto.d.ts (hunk 5 above)
Comment (Danielle9897) [reply src/Raven.Studio/typings/_studio/dto.d.ts]: removed - conflict was incorrectly reolved

File: src/Raven.Studio/typings/_studio/dto.d.ts (hunk 6 above)
Comment (Danielle9897) [reply src/Raven.Studio/typings/_studio/dto.d.ts]: removed

File: src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts (hunk 3 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts]: done

File: src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts]: done

File: src/Raven.Studio/wwwroot/App/views/database/tasks/exportDatabase.html (hunk 7 above)
Comment (Danielle9897) [reply src/Raven.Studio/wwwroot/App/views/database/tasks/exportDatabase.html]: done

File: src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html (hunk 9 above)
Comment (Danielle9897) [reply src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html]: Using text only

File: src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html (hunk 8 above)
Comment (Danielle9897) [reply src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html]: done

File: src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts:132]: done

File: src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts (hunk 2 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts]: done

File: src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts (hunk 1 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts]: done

File: src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html (hunk 9 above)
Comment (ml054) [reply src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html]: why not icon only? Like we do in other places or like github does?

Comment (Danielle9897) [reply src/Raven.Studio/wwwroot/App/views/database/tasks/importDatabaseFromFile.html]: ok - changed to icon only

File: src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts [hunk 10]
```diff
@@ -239,6 +231,42 @@ class exportDatabase extends viewModelBase {
...
+    getCommandTypeLabel(cmdType: commandLineType) {
+        return `Export Command - ${cmdType}`;
+    }
+
+    copyCommandToClipboard() {
+        const command = this.effectiveCommand();
+        copyToClipboard.copy(command, "Export command was copied to clipboard.");
+    }
+    
+    getCommand(commandType: commandLineType) {
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts:244]: purpose of extracting something into shared logic is to have *common* logic inside and custom *outside*

Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/tasks/exportDatabase.ts:244]: getCommand() should return [string, string]. First will be file name, second json. You don't need commandType arg here.

File: src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts [hunk 11]
```diff
@@ -72,51 +72,16 @@ class importDatabaseFromFile extends viewModelBase {
...
-            
-            return `curl -F 'importOptions=${json}' -F 'file=@${fileName}' ${commandEndpointUrl(db)}`;
-        });
-        
-        this.importCommandCmd = ko.pureComputed(() => {
-            const db = this.activeDatabase();
-            if (!db) {
-                return "";
+        this.effectiveCommand = ko.pureComputed(() => {
+            const cmdType = this.effectiveCommandType();
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts]: this method can be simplified to one line: ko.pureComputed(() => this.getCommand(this.effectiveCommandType()));

Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/tasks/importDatabaseFromFile.ts]: done

**Review by ml054 — APPROVED** [review]

**Pull request merged**